Nuvoton: M487 OTA support

cyliangtw opened the pull request:
<!--- Title -->

Description
-----------
This PR purpose is not to ask any qualification logo and not plan to process OCW.
It is one normal PR to support \ \amazon-freertos\vendors\vendor\boards\board\ports\ota for OTA-Agent job protocol. 

<!--- Describe your changes in detail -->

This PR add ports of M487 ota and it's verified by aws_test including FULL_OTA_AGENT_ENABLED, FULL_OTA_PAL_ENABLED and MQTTv4_ENABLED.
To run aws_test by cmake, the command line as: `$ cmake -DVENDOR=nuvoton -DBOARD=numaker_iot_m487_wifi -DCOMPILER=arm-keil -S. -Bbuild -G"Unix Makefiles" -DAFR_ENABLE_TESTS=1`
Attached the test log file for your reference.
[aws_test_log.txt](https://github.com/aws/amazon-freertos/files/5551167/aws_test_log.txt)


About loader part, we have one for customer reference and our chip has H/W secure boot capacity, it really depends on customer requirement. The loader’s major implementation is how to cooperate with OTA-client and secure boot could rely on M487 solution https://www.nuvoton.com/support/technical-support/technical-articles/TSNuvotonTechBlog-000086/.


Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

@yngki 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

cyliangtw committed:
M487 support OTA and upgrade aws_demo_eth Keil prj

cyliangtw committed:
Add M487 OTA part into CMakeLists

cyliangtw committed:
M487 aws-test support OTA

cyliangtw committed:
M487 OTA ensure file-close state before commit

cyliangtw committed:
M487 test runner & rtos heap config for OTA + MQTT test

yngki commented:
Conversation is occurring over email with the Nuvoton team.  PR will be updated in a few weeks with more details.

cyliangtw commented:
Hi @yngki ,
I just rebase and fixed 2 conflicts.
Also pass ota-demo including download ota image, reset-device & complete ota job.

yngki commented:
Thanks Chris, the OTA team will get back to you very soon on this.  My apologies for the delay.

pvyawaha commented on libraries/freertos_plus/aws/ota/include/aws_iot_ota_agent.h:
```diff
@@ -401,6 +401,9 @@ typedef struct
 typedef struct OTA_FileContext
 {
     uint8_t * pucFilePath; /*!< Local file pathname. */
+#if defined(__CC_ARM)
```
Can we enable this somewhere else like main.c instead of OTA library source?

cyliangtw replied:
@pvyawaha , to enable it in main.c can't be effective in the whole project and there are several ota common source code using aws_iot_ota_agent.h, such like as aws_iot_ota_agent.c, aws_iot_ota_mqtt.c.  More, this modification is for Keil compiler and not specific for some demo or some vendor. Refer to [Keil doc #pragma anon_unions](https://www.keil.com/support/man/docs/armcc/armcc_chr1359124984759.htm)

aggarw13 replied:
@cyliangtw, we would want to keep the OTA library platform independent, and instead add platform specific logic in config files. Can you please consider adding the pragma logic within `aws_ota_agent_config.h` and `FreeRTOSConfig.h`  files within the `vendors/nuvoton` directory?

cyliangtw replied:
@aggarw13 ,  thanks of your advisement and I moved to enable in nuvoton config files. So, I removed the previous commit of OTA lib and added one new commit of nuvoton config files.

pvyawaha commented on vendors/nuvoton/boards/numaker_iot_m487_wifi/aws_demos/config_files/aws_ota_agent_config.h:
```diff
@@ -0,0 +1,151 @@
+/**
+ * @brief The number of words allocated to the stack for the OTA agent.
+ */
+#define otaconfigSTACK_SIZE                     1500U //3584U
```
Cleanup or add comment on the commented value //3584

cyliangtw replied:
@pvyawaha , thanks of your suggestion. 
Fixed in commit e05644a .

pvyawaha commented on vendors/nuvoton/boards/numaker_iot_m487_wifi/aws_demos/config_files/aws_ota_agent_config.h:
```diff
@@ -0,0 +1,151 @@
+ * Enable data over HTTP - ( OTA_DATA_OVER_HTTP)
+ * Enable data over both MQTT & HTTP ( OTA_DATA_OVER_MQTT | OTA_DATA_OVER_HTTP )
+ */
+#define configENABLED_DATA_PROTOCOLS         ( OTA_DATA_OVER_MQTT )//( OTA_DATA_OVER_MQTT | OTA_DATA_OVER_HTTP )
```
OTA over HTTP will not be supported ? Cleanup and add comment

cyliangtw replied:
@pvyawaha , yes, M487 memory footprint is not suitable thru. HTTP. Added comment to guide thru. MQTT .

pvyawaha commented on vendors/nuvoton/boards/numaker_iot_m487_wifi/aws_demos/config_files/aws_ota_agent_config.h:
```diff
@@ -0,0 +1,151 @@
+ * This configurations parameter sets the maximum number of static data buffers used by
+ * the OTA agent for job and file data blocks received.
+ */
+#define otaconfigMAX_NUM_OTA_DATA_BUFFERS       2U
```
Ideally number of blocks requested should be same as number of buffers available , can this be increased to 4 by default?

cyliangtw replied:
@pvyawaha , M487 + freertos-plus-tcp can't afford 3 blocks, so to change both of `NUM_BLOCKS_REQUEST` and `NUM_OTA_DATA_BUFFERS` as 2 . Fixed in commit e05644a .

pvyawaha reviewed (changes requested):
changes requested for not changing the OTA library file for enabling anonymous unions.

cyliangtw committed:
Add comment and adjust aws_ota_agent_config to fulfill M487 platform memory footprint

cyliangtw committed:
Add comment and adjust \aws_tests\config_files\aws_ota_agent_config to fulfill M487 platform memory footprint

cyliangtw committed:
Keil project file support OTA demo thru. WiFi connectivity

cyliangtw commented:
Hi @pvyawaha , 
About to enable anonymous unions, it's compiler issue and can't resolve in main.c, so still to keep the change of OTA lib. 
Other issues are fixed in the 3 commits or explained.
Is there any concern or comment ?

aggarw13 commented:
/bot run checks

aggarw13 commented on vendors/nuvoton/boards/numaker_iot_m487_wifi/CMakeLists.txt:
```diff
@@ -149,6 +150,30 @@ target_sources(
+target_include_directories(
+    AFR::ota::mcu_port
+    INTERFACE
+            "${AFR_ROOT_DIR}/libraries/abstractions/pkcs11/corePKCS11/source/portable/mbedtls/include"
```
@cyliangtw, this path is incorrect for the latest commit of `corePKCS11` library on the master branch which is causing CI build check failures. Can you please make sure to update the submodule in your local workspace and update this path?

cyliangtw replied:
@aggarw13 ,  yes, this folder was removed in corePKCS11#13224d85 on 2020/12/10. 
Fixed CMakeLists.txt in the new commit 4b9bfde and the new attached log shows all pass. 
[ota_test_v202101.txt](https://github.com/aws/amazon-freertos/files/5901705/ota_test_v202101.txt)

aggarw13 commented on vendors/nuvoton/boards/numaker_iot_m487_wifi/CMakeLists.txt:
```diff
@@ -149,6 +150,30 @@ target_sources(
+    INTERFACE
+            "${AFR_ROOT_DIR}/libraries/abstractions/pkcs11/corePKCS11/source/portable/mbedtls/include"
+            "${AFR_ROOT_DIR}/libraries/abstractions/pkcs11/corePKCS11/source/include"
+            "${AFR_ROOT_DIR}//libraries/3rdparty/pkcs11"
```
Seems like an extra slash has been added which can be removed
```suggestion
            "${AFR_ROOT_DIR}/libraries/3rdparty/pkcs11"
```

cyliangtw replied:
@aggarw13 , Thanks of your notification. It's fixed in the new commit 4b9bfde.

cyliangtw committed:
Update include path to fulfill the latest commit of corePKCS11 lib

aggarw13 commented:
/bot run checks

aggarw13 commented:
/bot run checks

aggarw13 commented:
> Hi @pvyawaha , 
About to enable anonymous unions, it's compiler issue and can't resolve in main.c, so still to keep the change of OTA lib. 
Other issues are fixed in the 3 commits or explained.
Is there any concern or comment ?

@cyliangtw, as mentioned in the [reply here](https://github.com/aws/amazon-freertos/pull/2754/files#r568844219), would you be able to add the compiler specific logic in config files?

cyliangtw committed:
Enable anon_unions for Keil compiler in nuvoton config files

cyliangtw commented:
> > Hi @pvyawaha ,
> > About to enable anonymous unions, it's compiler issue and can't resolve in main.c, so still to keep the change of OTA lib.
> > Other issues are fixed in the 3 commits or explained.
> > Is there any concern or comment ?
> 
> @cyliangtw, as mentioned in the [reply here](https://github.com/aws/amazon-freertos/pull/2754/files#r568844219), would you be able to add the compiler specific logic in config files?

@aggarw13 ,  yes, I already added the compiler specific logic in nuvoton config files at commit bebe57d and removed the unnecessary  commit of `libraries/freertos_plus/aws/ota/include/aws_iot_ota_agent.h` .

aggarw13 commented:
/bot run checks

aggarw13 commented:
/bot run checks

cyliangtw commented:
Hi @aggarw13 ,
  The build checks failure is not true, the details show this PR all pass as below, please check it.
```
Artifacts of PR Build Check
Date Built: 2021-02-03 10:07:51

Fail Count: 0

Pass Count: 54

Unknown Status Count: 0

Return to Pull Request: https://github.com/aws/amazon-freertos/pull/2754
```

yourslab commented:
/bot run checks

yourslab commented:
/bot run checks
